_posts/2016-04-10-running-multiple-apps-in-umbrella-project.md

```diff
@@ -0,0 +1,114 @@
+
+As NectarCommerce extensions are envisioned to use with Nectar and would be dependent on Nectar directly or indirectly, we chose umbrella project to have Nectar, extensions\_manager, user\_app, and extensions apps as part of it.
+
+We plan to use `phoenix app` for extensions where web-component is needed and which would be part of umbrella project.
+
+The `user_app` would be a resting ground where Nectar and extensions are evaluated for compile time requirements and prepared to be used together.
+
+As of now, `user_app` has following responsibilities:
+
+- [defining nectar as umbrella dependency](https://github.com/vinsol/nectarcommerce/pull/47/files#diff-cfadcd558b7d7d0fa0bb1b43ff49ba98R44)
+- [defining available extensions as umbrella dependencies](https://github.com/vinsol/nectarcommerce/pull/47/files#diff-cfadcd558b7d7d0fa0bb1b43ff49ba98R45)
+- [acting as a pass-through for all the outbound requests](https://github.com/vinsol/nectarcommerce/pull/47/files#diff-02eb191c0680951b40ba159f4163a7c6R21)
+
+## Running Multiple Phoenix Applications together in an umbrella ##
+
+Phoenix applications are in the end just another OTP application, with a few modifications and configuration tweaks we can expect them to run together in same umbrella project. We need to make sure that
+
+1. They share the same configuration for database, while not mandatory if all the applications are utilizing the same databases, keeping the configuration in one place is a good practice. We can create a ```shared_config.exs``` at root of umbrella with some code like this:
+	
+	
+	```elixir
+	config :shared, :db_config,
+	  username: "shared_username",
+	  pool_size: 200
+	```
+	
+	And, then in each individual application at the end of file:
+	
+	```elixir
+	shared_db_config = Mix.Config.read!("../../shared_config.exs")[:shared][:db_config]
+	config :app, App.Repo, shared_db_config
+	```
+	
+	Obviously this is use case dependent some applications may not share a database and only rely on each other for method calls.
+
+2. Starting the phoenix server. If we look closely at the application file for our phoenix app, we can see a supervisor for endpoint, this starts the phoenix server and binds it to the port.
+
+   We can either start the application on different ports and use an external server like nginx to dispatch on appropriate port based on business requirements.
+
+   We can also use forwards at the end of route files, using the ```forward "/", App.router```. from one of the application and start the server from there.
+
+   If we go with the **latter approach** we need to [make sure that only the endpoint supervisor of current application is in supervision tree](https://github.com/vinsol/nectarcommerce/pull/47/files#diff-001ce116edeff3304fe941b84c128c5dR12). Else we will get an address already in use error.
```

Comment:
If we go with the latter approach we need to make sure that only the endpoint supervisor of current application is in supervision tree, else we will get an 'address already in use' error.